# Description

This docker image tries to summarize an essential toolset for NGS data secondary analysis, that typically consists of the stages:

* QC
* Pre-alignment
* Alignment
* Post-align
* Variant calling
* Filtering and annotation
* Summary reporting

While this docker image does not provide analysis scripts, it exposes appropriate applications, that can be used to be built pipeline.

# Configuration

All the tools, listed below, are installed into `/opt/` directory.

For each of the tools - two environment variables are available:

* `<TOOL_NAME>_HOME` - points to the installation location of the corresponding tool. E.g. for bedtools: `BEDTOOLS_HOME=/opt/bedtools`
* `<TOOL_NAME>_BIN` - points to the executable of the corresponding tool. E.g. for bedtools: `BEDTOOLS_BIN=/opt/bedtools/bin/bedtools`

These variables can be used in scripts to launch the applications, instead of hardcoding tools version/locations.

# General purpose tools

* R 3.2
* Java 7/8
* Python 2.7

# NGS tools

* `STAR` (https://github.com/alexdobin/STAR): Spliced Transcripts Alignment to a Reference
* `ngs-bits` (https://github.com/imgag/ngs-bits): Short-read sequencing tools
* `cufflinks` (http://cole-trapnell-lab.github.io/cufflinks): Cufflinks assembles transcripts, estimates their abundances, and tests for differential expression and regulation in RNA-Seq samples
* `Subread` (https://sourceforge.net/projects/subread/): Subread aligner can be used to align both gDNA-seq and RNA-seq reads. Subjunc aligner was specified designed for the detection of exon-exon junction. For the mapping of RNA-seq reads, Subread performs local alignments and Subjunc performs global alignments
* `samtools` (https://github.com/samtools/samtools): A flexible generic format for storing nucleotide sequence alignment. SAMtools provide efficient utilities on manipulating alignments in the SAM format
* `picard` (https://github.com/broadinstitute/picard): A set of Java command line tools for manipulating high-throughput sequencing (HTS) data and formats
* `rnaseqc` (https://software.broadinstitute.org/cancer/cga/rna-seqc): A java program which computes a series of quality control metrics for RNA-seq data
* `bedtools` (https://github.com/arq5x/bedtools2): Collectively, the bedtools utilities are a swiss-army knife of tools for a wide-range of genomics analysis tasks
* `bwa` (https://sourceforge.net/projects/bio-bwa/): Aligning sequencing reads against a large reference genome
* `snpEff` (https://sourceforge.net/projects/snpeff/): A fast variant effect predictor (SNP, MNP and InDels) for genomic data
* `VarDictJava` (https://github.com/AstraZeneca-NGS/VarDictJava): A variant discovery program written in Java and Perl
* `abra` (https://github.com/mozack/abra): A realigner for next generation sequencing data. It uses localized assembly and global realignment to align reads more accurately, thus improving downstream analysis (detection of indels and complex variants in particular)
* `abra2` (https://github.com/mozack/abra2): An updated implementation of ABRA featuring: RNA support,Improved scalability, Improved accuracy, Improved stability and usability 
* `mutect` (https://software.broadinstitute.org/cancer/cga/mutect): A method developed at the Broad Institute for the reliable and accurate identification of somatic point mutations in next generation sequencing data of cancer genomes
* `scalpel` (https://sourceforge.net/projects/scalpel/): Bioinformatics pipeline for discovery of genetic variants from NGS reads
* `bowtie2` (https://sourceforge.net/projects/bowtie-bio/): An ultrafast, memory-efficient short read aligner for short DNA sequences from next-gen sequencers
* `STAR-Fusion` (https://github.com/STAR-Fusion/STAR-Fusion): STAR-Fusion uses the STAR aligner to identify candidate fusion transcripts supported by Illumina reads. STAR-Fusion further processes the output generated by the STAR aligner to map junction reads and spanning reads to a reference annotation set
* `bamtools` (https://github.com/pezmaster31/bamtools): C++ API & command-line toolkit for working with BAM data
* `FastQC` (https://github.com/s-andrews/FastQC): Quality Control application for FastQ files
* `Trimmomatic` (http://www.usadellab.org/cms/?page=trimmomatic): Flexible read trimming tool for Illumina NGS data
